src/api_manager/service_control/aggregated.cc

```diff
@@ -441,61 +483,70 @@ Status Aggregated::GetStatistics(Statistics* esp_stat) const {
   return Status::OK;
 }
 
-template <class RequestType>
-const std::string& Aggregated::GetApiRequestUrl() {
-  if (typeid(RequestType) == typeid(CheckRequest)) {
-    return url_.check_url();
-  } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
-    return url_.quota_url();
-  } else {
-    return url_.report_url();
-  }
+template <>
+const std::string& Aggregated::GetApiRequestUrl<CheckRequest>() {
+  return url_.check_url();
+}
+template <>
+const std::string& Aggregated::GetApiRequestUrl<ReportRequest>() {
+  return url_.report_url();
+}
+template <>
+const std::string& Aggregated::GetApiRequestUrl<AllocateQuotaRequest>() {
+  return url_.quota_url();
 }
 
-template <class RequestType>
-int Aggregated::GetHttpRequestTimeout() {
-  int timeout_ms = 0;
+template <>
+int Aggregated::GetHttpRequestTimeout<CheckRequest>() {
+  return check_timeout_ms_;
+}
+template <>
+int Aggregated::GetHttpRequestTimeout<ReportRequest>() {
+  return report_timeout_ms_;
+}
+template <>
+int Aggregated::GetHttpRequestTimeout<AllocateQuotaRequest>() {
+  return quota_timeout_ms_;
+}
 
-  // Set timeout on the request if it was so configured.
-  if (typeid(RequestType) == typeid(CheckRequest)) {
-    timeout_ms = kCheckDefaultTimeoutInMs;
-  } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
-    timeout_ms = kAllocateQuotaDefaultTimeoutInMs;
+template <>
+int Aggregated::GetHttpRequestRetries<CheckRequest>() {
+  return check_retries_;
+}
+template <>
+int Aggregated::GetHttpRequestRetries<ReportRequest>() {
+  return report_retries_;
+}
+template <>
+int Aggregated::GetHttpRequestRetries<AllocateQuotaRequest>() {
+  return quota_retries_;
+}
+
+template <>
+const std::string& Aggregated::GetAuthToken<CheckRequest>() {
+  if (sa_token_) {
+    return sa_token_->GetAuthToken(
+        auth::ServiceAccountToken::JWT_TOKEN_FOR_SERVICE_CONTROL);
   } else {
-    timeout_ms = kReportDefaultTimeoutInMs;
+    static std::string empty;
```

Comment:
use pointer with new since std::string is trivially destructible?

Reply:
hmm maybe empty global string is another way to go since empty string is used in multiple times in this file.

Reply:
Thanks.  Updated.